broker/server_apis.go

```diff
@@ -0,0 +1,246 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ***************************************************************/
+
+package broker
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/pelicanplatform/pelican/param"
+	"github.com/pelicanplatform/pelican/token_scopes"
+	log "github.com/sirupsen/logrus"
+)
+
+type (
+
+	// Structure for an origin's POST to the broker
+	originRequest struct {
+		Origin string `json:"origin"`
+		Prefix string `json:"prefix"`
+	}
+
+	// Structure for an origin retrieval response from the broker
+	originResp struct {
+		Status  string          `json:"status"`
+		Request reversalRequest `json:"request"`
+	}
+
+	// Error response for a request or retrieval
+	brokerErrResp struct {
+		Status string `json:"status"`
+		Msg    string `json:"msg"`
+	}
+
+	// Structure for an origin calling back to the cache
+	callbackRequest struct {
+		RequestId string `json:"request_id"`
+	}
+)
+
+func retrieveRequest(ctx context.Context, ginCtx *gin.Context) {
+	timeoutStr := "5s"
+	if val := ginCtx.Request.Header.Get("X-Pelican-Timeout"); val != "" {
+		timeoutStr = val
+	}
+
+	timeoutVal, err := time.ParseDuration(timeoutStr)
+	if err != nil {
+		ginCtx.String(http.StatusBadRequest, "Failed to parse X-Pelican-Timeout header to a duration (example: 5s)")
+		ginCtx.Abort()
+		return
+	}
+
+	originReq := originRequest{}
+	if err := ginCtx.Bind(&originReq); err != nil {
+		ginCtx.String(http.StatusBadRequest, "Failed to parse the origin's retrieve request")
+		ginCtx.Abort()
+		return
+	}
+
+	token := ginCtx.Request.Header.Get("Authorization")
+	token, hasPrefix := strings.CutPrefix(token, "Bearer ")
+	if !hasPrefix {
+		ginCtx.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"status": "error", "msg": "Bearer authorization required for callback"})
+		return
+	}
+
+	ok, err := verifyToken(ctx, token, originReq.Prefix, param.Server_ExternalWebUrl.GetString(), token_scopes.Broker_Retrieve)
+	if err != nil {
+		log.Errorln("Failed to verify token for reverse request:", err)
+		ginCtx.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"status": "error", "msg": "Failed to verify provided token"})
+		return
+	}
+	if !ok {
+		ginCtx.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"status": "error", "msg": "Authorization denied"})
+	}
+
+	req, err := handleRetrieve(ctx, ginCtx, originReq.Prefix, originReq.Origin, timeoutVal)
+	if errors.Is(err, errRetrieveTimeout) {
+		ginCtx.JSON(http.StatusOK, gin.H{"status": "timeout", "request": gin.H{}})
+		return
+	} else if err != nil {
+		ginCtx.JSON(http.StatusInternalServerError, brokerErrResp{Status: "error", Msg: "Failure when retrieving requests for this origin"})
+		return
```

Comment:
Since we do have a struct to represent error message for broker, let's utilize this struct and update all `gin.H{}` instances.

Reply:
I would also be nice to have enums for `Status`

Reply:
Good ideas.  In commit a6b121c, I refactored to use a common structure throughout for all broker responses from this function.